File: 198-house-robber/house-robber.java
class Solution {
    public int rob(int[] nums) {
       

    //  int[] dp = new int[nums.length+1];
    // dp[0] = nums[0];
    int prev = nums[0];
    int prev2 = 0;
    int ans=prev;


    int pick,nonpick;
    for (int i = 1;i<nums.length;i++){
         nonpick = prev;
        
        pick = prev2 + nums[i];

        ans = Math.max(pick,nonpick);
        prev2 = nonpick;
        prev = ans;
        

    }

    return ans;



        // int[] dp = new int[nums.length];
        // Arrays.fill(dp,-1);
        // return helper(nums,nums.length-1,dp);
        
    }

    // int helper(int[] nums,int i,int[] dp){
     
    //  if (i == 0) return nums[0];

    //  if (i <0 ) return 0;

    // if (dp[i] != -1) return dp[i];

    //  int pick = helper(nums,i-2,dp) + nums[i];
    //  int nonpick = helper(nums,i-1,dp);

    //  dp[i] = Math.max(pick,nonpick);

    //  return dp[i];



    // }
}

File: 198-house-robber/house-robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        
        dp = [-1]*len(nums)
        return self.helper(0, nums, 0, dp)

    def helper(self, ind, nums, money, dp):


        if ind >= len(nums):
            return 0
            
        if dp[ind] != -1:
            return dp[ind]
            
        rob = nums[ind] + self.helper(ind + 2, nums, money, dp)

        no_rob = self.helper(ind + 1, nums, money, dp)

        dp[ind] = max(rob, no_rob)
        return dp[ind]
        #return max(rob, no_rob)

